### DISCUSSION TITLE
What is the value of using RenderMode.InteractiveServer"?

### DISCUSSION THREAD

**Discussion by DavidThielen** [dotnet/aspnetcore#53201] [Q&A] [answered]
Hi;

I don't understand the value of this. And the Blazor team is smart which means this is something worth the effort. So I'd like to understand the value in it so I use it when appropriate.

At present I am using (code first created in .NET 6, now using .NET 8):
```razor
<component type="typeof(HeadOutlet)" render-mode="Server" />
```
But now the recommended approach in .NET 8 is:
```razor
<Routes @rendermode="@RenderMode.InteractiveServer"/> 
```
Which I assume can be done as:
```razor
<component type="typeof(HeadOutlet)" render-mode="ServerPrerendered" />
```
So here's my question. When I switch to `ServerPrerendered`, it then calls `OnInitializedAsync()` twice. For a lot of my pages the performance hit is reading the appropriate data from the database. Reading the DB twice means it will take twice as long to load the page.

Where's the win in this? I could see a win if there was a new `OnFastInitialized()` (no Async version), followed by `OnInitializedAsync()`. But we don't have that. So as best as I can tell, all `ServerPrerendered` delivers is slower page initialization.

And an additional negative for `ServerPrerendered` - on the second call to `OnInitializedAsync()` all the cascading parameters are null. This is a giant problem for my code. So significant, that even if `ServerPrerendered` had some major advantages, I still probably could not use it.

So... what is the advantage of `ServerPrerendered` that makes it worth the slower page load and loss of cascading parameters?

thanks - dave

**Comment by javiercn** [comment, marked as answer]
The difference is that in one case you are prerendering and in the other you aren't. The benefit of prerendering is that the content is available to the user much faster. 
* In the prerendered case, the content arrives in the initial response.
* In the non-prerendered case, you first need to send a request, get a response, then start a SignalR connection (a couple of requests), then trigger the render.
  * That's likely between 2 and 4 times slower depending on the bandwidth and can be significant depending on what drives the time to render the UI on the screen.

There are strategies you can follow to avoid hitting the database twice:
* You can add a cache at the server level for the data you are using.
* You can persist the prerendered component state to avoid the double lookup.
* https://learn.microsoft.com/en-us/aspnet/core/blazor/components/prerender?view=aspnetcore-8.0#persist-prerendered-state

**Reply by DavidThielen** [reply]
First off, thanks for answering.

Reading your answer, I'm still not getting it. Lets take my three main use cases, which I think hold for the vast majority of business apps:
1. The user's dashboard displays items relevant to them. Every user will have a different list.
2. A list of search results. This list is different for every search.
3. A page to Create, Read, or Update a complex record. The data is unique to the record requested.

For each of the above, the page is a useless empty skeleton until the data is read in. Getting that empty skeleton up quicker is of no utility. All that matters is how quickly the populated page is displayed. And as best as I can tell pre-rendering is of no help there. Am I missing something about this?

In addition, for each of the above, the data is unique to the user and what parameters they selected. So there is no way to cache the data, because each page is displaying data that has likely never been selected in the previous couple of hours (or ever). Or do you mean read in pre-rendering and cache it for the final rendering? If so, how is pre-rendering of any value here?

Persisting the pre-rendered state to the final rendering makes sense. I can see how that saves some time. But that doesn't answer why have a pre-render step. It merely says if you have the two steps, here's how to reduce the delay the double steps cause. But if I don't set to `InteractiveServer`, this issue goes away.

So I'm back at my original question - if the page is worthless without the data, what is the advantage, if any, of `InteractiveServer`?

I want to stress, I'm not trying to be difficult. A number of smart people put in a lot of time to provide this render mode. I want to understand why/when it makes sense to use it. So that I will use it when appropriate.

thanks - dave

**Reply by javiercn** [reply]
@DavidThielen seems that in your case you don't care about how "fast" the data is displayed, but this is an important concern in many apps/sites. To exaggerate this a bit, let's say it took 30secs / 1 minute to get the data displayed on the screen. Are you ok with the users staring for 30s/1m at a blank/loading screen while the app boots? That is why prerendering is important. If this is not a big deal for you (because your users are "forced" to use your app) then simply disable it. If otherwise, your users might give up, then having content on the screen (as soon as possible) while the rest of the data loads, it's important.

**Reply by DavidThielen** [reply]
@javiercn I want to display the page very quickly to users. But the page is of no value without the data.

Look at it this way. I do a search on Google. If Google instantly put up a "results" page with no search results, then populated the page with the search results, that fir no results page is of zero use. No user is going to be happier that they first got a page with no results.

That's my question. Normally I can render a page, with data, in under 1/2 second. So I think that's a good user experience. But I don't see the advantage of putting the empty page up in 0.05 seconds, followed by the populated page 0.5 seconds later vs. the populated page in 0.5 or even 1.5 seconds.

And I'm struggling to think of a use case where an unpopulated page is of interest to a user. Or event to a search crawler as the interesting words for a search tend to be the data populating the page.

Can you give me a use case where this is an advantage? TIA

**Reply by javiercn** [reply]
> Look at it this way. I do a search on Google. If Google instantly put up a "results" page with no search results, then populated the page with the search results, that fir no results page is of zero use. No user is going to be happier that they first got a page with no results.

Displaying a skeleton and a loading icon is better than waiting on an empty page, at which point many users simply give up.

> That's my question. Normally I can render a page, with data, in under 1/2 second. So I think that's a good user experience. But I don't see the advantage of putting the empty page up in 0.05 seconds, followed by the populated page 0.5 seconds later vs. the populated page in 0.5 or even 1.5 seconds.

If according to your goals your site is simple/fast enough for your users without prerendering, it's ok to disable it. The default guidance based on data is that it's normally beneficial to prerender on the server to provide a better user experience.

In your case, if you are developing an app for "internal customers" you might not benefit from it and it's ok to disable it in that case, but in general, that's the exception not the rule.

For more info check:
https://developers.google.com/search/docs/appearance/core-web-vitals
https://www.cloudflare.com/learning/performance/more/website-performance-conversion-rates/

Also, to add an additional comment to your point on being able to render a page with data in under 1/2 second. It matters a lot how you are measuring this, and how your customers are accessing the site.

If you have customers with high bandwidth and low latency, it probably won't make such a big difference. However, that's not an assumption we generally make (the opposite is more the case, your app needs to work "reasonably well" even on a 3G connection in the middle of nowhere). 

So, things like latency, affect the experience much more (very different to have a 10ms RTT vs a 200ms RTT) where if you have a chain of at least 4 requests until users can see data, we are talking about a difference between 40ms and 800ms (just counting on the data going back and forth).

Add to that the time it takes to transfer data to the client (scripts, etc.) which if your connection is limited and not very stable (mobile) will introduce further delays.

It's not super clear to me how you reconcile the "issue" of hitting the database twice with the statement that you can load and render the data very fast. In my reasoning, if your queries are not expensive, it doesn't matter that you hit the DB twice (first on prerendering/interactive) and if the queries are expensive, it likely means that they take time to run, at which point the app will benefit from prerendering by allowing other resources to load in parallel while the data is being fetched.

Again, this is all up to your specific app/user scenarios. If you work in a domain where users have strong internet connection and there's no negative impact on wait times (they won't leave your site) then it might not be worth for you to use prerendering. 

In pretty much any other case, delivering the content as fast as possible to users under any environment is an important business requirement.

**Reply by DavidThielen** [reply]
@javiercn That's a good point about the low bandwidth user - I can have everything ready in 0.2 seconds, but if it's a lot of content, then they're going to have the page render slowly. Thanks for walking me through this - I know I asked lots of questions. I appreciate your guidance.

thanks - dave

**Reply by Tony20221** [reply]
@DavidThielen  Having a progress bar with a skeleton page with menus and such is better than just a progress bar on a blank page which is better than a blank page with nothing on it. A progress bar that actually displays the status of remaining time is better than a dummy animated gif.

In the prerender page, you can display some text that contains some information to grab the users' attention while the page is waiting for the database data. 
It can be the latest news from your department. It could be about the next downtime window for maintenance work. Some kind of announcement. Or just say the database is experiencing heavy load and the page will load in less than x seconds. The goal is to lessen the feeling of waiting.

**Reply by DavidThielen** [reply]
@Tony20221 My fundamental question was not why have 2 passes. I understood that. My question was why call `OnInitializedAsync()` twice. Why not instead have `OnFirstPassAsync()` and `OnSecondPassAsync()` and the first pass is everything that can be done quickly to render the static skeleton page and the second pass is everything including the DB calls.

Even after all the discussion above, I still don't understand what they didn't break this out into two distinct calls.

**Reply by Tony20221** [reply]
Is the issue you're encountering not knowing whether you're in the first run or the second run to decide when to retrieve data from the database in the OnInitializedAsync method?

**Reply by DavidThielen** [reply]
@Tony20221 Yes. So I can then not hit the DB on the first pass. And yes, there are tricks you can pull at present to figure this out (Cascading Parameter == null), but those are not guaranteed to work in future versions.
